File: 1593-split-a-string-into-the-max-number-of-unique-substrings/1593-split-a-string-into-the-max-number-of-unique-substrings.cpp
class Solution {
public:
    int ans = 0;
    unordered_set<string> st;
    void solve(string &s, int i, string t) {
        if (i == s.size()) {
            ans = max(ans, (int)st.size());
            return;
        }
        for (int j = i; j < s.size(); j++) {
            t += s[j];
            if (st.find(t) == st.end()) {
                st.insert(t);
                solve(s, j+1, "");
                st.erase(t);
            }
        }
    }
    int maxUniqueSplit(string s) {
        solve(s, 0, "");
        return ans;
    }
};

File: 1593-split-a-string-into-the-max-number-of-unique-substrings/1593-split-a-string-into-the-max-number-of-unique-substrings.java
class Solution {
    private int ans = 0;
    Set<String> st = new HashSet<>();
    void solve(String s, int i, String t) {
        if (i == s.length()) {
            ans = Math.max(ans, st.size());
            return;
        }
        for (int j = i; j < s.length(); j++) {
            t += s.charAt(j);
            if (!st.contains(t)) {
                st.add(t);
                solve(s, j+1, "");
                st.remove(t);
            }
        }
    }
    public int maxUniqueSplit(String s) {
        solve(s, 0, "");
        return ans;
    }
}